### DISCUSSION TITLE
How to use numref for tables generated in code-cells?

### DISCUSSION THREAD

**Discussion by cnjr2** [executablebooks/meta#1190] [Q&A] [answered]
**TLDR**: Is it possible to get table numbering to behave in the same way was for figures when working with `{code-cell}`s? (Related to issues #72 and #243)

**In mode detail:**

It would be great to have this:

````
```{code-cell} ipython3
---
mystnb:
  table:   # <--- This would be great!
    caption: |
      Hey everyone its **party** time!
    caption_before: True
    name: fun-fish-table
---
import pandas as pd
pd.DataFrame({"qwd": [1, 10], "wefwef": [10, 20]})
```
````

This is a reference to `{numref}'fun-fish-table'` rendering as `Tab. 1`.

For figures this works, based on discussions in issues [#72](https://github.com/executablebooks/MyST-NB/issues/72) and [#189](https://github.com/executablebooks/MyST-NB/issues/189), a new way to provide context when rendering figures was added with PR [#243](https://github.com/executablebooks/MyST-NB/pull/243) and is described in the [docs](https://jupyterbook.org/en/stable/content/code-outputs.html#images))

````
```{code-cell} ipython3
---
mystnb:
  image:
    width: 200px
    alt: fun-fish
    classes: shadow bg-primary
  figure:
    caption: |
      Hey everyone its **party** time!
    name: fun-fish
---
from IPython.display import Image
Image("../images/fun-fish.png")
```
````

This is great, as it works perfectly in conjunction with `{numref}'fun-fish-table'` where it tells me `Fig. 1`.

But for tables it seems different. I know it does work with for example:

````
```{list-table} This table title
:header-rows: 1
:name: fun-list-table

* - Training
  - Validation
* - 0
  - 5
* - 13720
  - 2744
```
````

I can refer to the table with `{numref}'fun-list-table'` and get it to render as `Tab. 1`.

Now when I do the following (from the [docs](https://myst-nb.readthedocs.io/en/latest/render/format_code_cells.html#images-and-figures)):

````
```{code-cell} ipython3
---
mystnb:
  figure:
    align: center
    caption_before: true
    caption: This is my table caption, above the table
    name: another-table # NOTE: added
---
import pandas
df = pandas.DataFrame({"column 1": [1, 2, 3]})
df = df.style.set_table_attributes('class="dataframe align-center"')
df
```
````

I get `Fig. 1` instead of `Tab. 1` when referring to it with `{numref}'another-table'`.

Is there a way to align figure and table behavior, or would one need to get into [custom rendering](https://myst-nb.readthedocs.io/en/latest/render/format_code_cells.html#customise-the-render-process)?

**Comment by cnjr2** [comment]
Is there perhaps a way to keep a counter across a book to somehow use to number tables?

**Comment by cnjr2** [comment]
Is this an awkward feature to implement, or just not useful @chrisjsewell ?

**Comment by cnjr2** [comment, marked as answer]
This is ultimately my workaround for this:

````
```{code-cell} python
---
tags: [remove-input]
mystnb:
  markdown_format: myst
---
from IPython.display import display, Markdown
import pandas as pd

display(
  Markdown(
    "\n".join(
      [
        "```{table} This is markdown table with caption.",
        ":name: markdown_table",
        # ":width: 50%",
        pd.DataFrame({"x": [1, 2, 3, 4], "y": [5, 2, 7, 9]}).to_markdown(),
        "```"
      ]
    )
  )
)
```
````

I think the option below would be nice as it would align with the figures:

````
```{code-cell} python
---
mystnb:
  table:
    caption: This is markdown table with caption.
    caption_before: True
    name: markdown_table
---
import pandas as pd
pd.DataFrame({"x": [1, 2, 3, 4], "y": [5, 2, 7, 9]})
```
````
